### ISSUE TITLE
Dataplane

### ISSUE THREAD

**Issue by jcp19** [viperproject/VerifiedSCION#169]
TODO:
- [x] bfdSend.Mem
- [x] bfdSend.String
- [x] bfdSend.Send (skipped)
- [x] Strengthen spec of decodeLayers
- processPkt and dependencies
  - [x] reset
  - cases:
    - [x] interBFD
    - [x] intraBFD
    - [x] onehop (#191)
    - [x] scion
        - [x] resolveInbound
        - [x] connect with the rest of the proof
- Run
  - [x] realized that the spec for BatchConn is unnecessarily restrictive, we should make it weaker - use wildcard perms instead of full perms
  - [x] converge in postconditions of `processPkt`
  - [x] merge #240 
  - [x] verify calls to the closures in `(*DataPlane).Run(...)` (#244)
  - Packscmp (#126, #192) (partially done in #243, but we might have run into a bug that keeps us from verifying it further)
      - [x] come up with a spec for `packSCMP` (#243)
      - [x] verify it according to the given spec (#243)
      - [ ] remove all `@TODO()` that precede calls to it (blocked on a bug from scion)

In summary, we verified mem. safety for when receiving and sending SCION packets (except when we send SCMP packets as reply - ongoing and depends on `packSCMP`), when handling `empty` and when handling `onehop` packets.
This means that the only things missing are the `epic` paths and the generation of `SCMP` packets.

## Nice to have:
On a first step:
  - [ ] processEpic (#196,  #261) - kinda done, but we need to debug performance problems
      - [x] libepic (#199, #266)
  - newBFDSend (and the setter methods that call this)
    - requires making slayers.SCION.Mem() more general (to allow nil ubufs)
    - setters of bfd connections in the dataplane
        - AddExternalInterfaceBFD
        - AddNextHopBFD
    - send and new
  - missing getters and setter

Later:
  - finish `prepareSCMP` (~~#245~~, #360) past the `@TODO()`
      - this may require some refactor of `(*SCION).Mem` and the introduction of a ghost field to describe whether a SCION struct was created manually or parsed from a slice
  - Split memory footprint predicates from functional properties, as started in https://github.com/viperproject/VerifiedSCION/pull/228
  - ~~finish and merge the refactor (#122, #155)~~
      - adapt according to https://github.com/viperproject/gobra/pull/619
  - ~~merge the latest commits in the upstream SCION (#190)~~
  - ~~check op. for linearity to use our linear type system~~
  - allow `_` perm in iterated collections with range (gobra)
  - bfdSend
  - ~~[x] issue #2~~
  - address all files that have been specified but not verified
  - ~~adapt to the new map encoding (https://github.com/viperproject/gobra/pull/606)~~
  - dataplane adapter
  - ~~ghost fields to simplify Dataplane~~

**Comment by jcp19** [comment]
I will close this issue. All major tasks that are still missing have open issues already, we will track progress through them.
